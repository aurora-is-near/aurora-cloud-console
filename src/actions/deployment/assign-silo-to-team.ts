"use server"

import { getTeamOnboardingForm } from "@/actions/onboarding/get-onboarding-form"
import { getUnassignedSiloId } from "@/actions/silos/get-unassigned-silo-id"
import { BASE_TOKENS } from "@/constants/base-token"
import { createAdminSupabaseClient } from "@/supabase/create-admin-supabase-client"
import { Silo } from "@/types/types"
import {
  assertNonNullSupabaseResult,
  assertValidSupabaseResult,
} from "@/utils/supabase"

export const assignSiloToTeam = async (
  teamId: number,
): Promise<Silo | null> => {
  const supabase = createAdminSupabaseClient()
  const onboardingForm = await getTeamOnboardingForm(teamId)

  if (!onboardingForm) {
    throw new Error(`Onboarding form not found for team ${teamId}`)
  }

  const { chainName, baseToken } = onboardingForm

  // 1. Get next available unassigned silo
  const unassignedSiloId = await getUnassignedSiloId()

  // Exit early if no unassigned silos, found
  if (!unassignedSiloId) {
    return null
  }

  // 2. Assign silo to a team
  await supabase
    .from("silos_teams")
    .insert([{ team_id: teamId, silo_id: unassignedSiloId }])

  // 3. Update the silo based on the onboarding form details
  const siloUpdateData: Partial<Silo> = {}

  if (chainName) {
    siloUpdateData.name = chainName
  }

  const baseTokenConfig = BASE_TOKENS[baseToken]

  if (baseTokenConfig) {
    siloUpdateData.base_token_name = baseTokenConfig.name
    siloUpdateData.base_token_symbol = baseToken
  }

  const updateSiloResult = await supabase
    .from("silos")
    .update(siloUpdateData)
    .eq("id", unassignedSiloId)
    .select()
    .single()

  assertNonNullSupabaseResult(updateSiloResult)
  assertValidSupabaseResult(updateSiloResult)

  return updateSiloResult.data
}
