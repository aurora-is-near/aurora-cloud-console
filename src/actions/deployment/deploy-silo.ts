"use server"

import { getTeamOnboardingForm } from "@/actions/onboarding/get-onboarding-form"
import { updateTeam } from "@/actions/teams/update-team"
import { BASE_TOKENS } from "@/constants/base-tokens"
import { createAdminSupabaseClient } from "@/supabase/create-admin-supabase-client"
import { OnboardingForm, Silo } from "@/types/types"
import { createSiloDeployerApiClient } from "@/utils/silo-deployer-api/silo-deployer-api-client"
import {
  assertNonNullSupabaseResult,
  assertValidSupabaseResult,
} from "@/utils/supabase"

const assignSiloToTeam = async (
  teamId: number,
  { baseToken, chainName }: OnboardingForm,
): Promise<Silo | null> => {
  const supabase = createAdminSupabaseClient()

  // 1. Get next available unassigned silo
  const unassignedSiloResult = await supabase
    .from("silos")
    .select("id, teams(id)")
    .is("teams", null)
    .order("id", { ascending: true })
    .limit(1)
    .maybeSingle()

  assertValidSupabaseResult(unassignedSiloResult)

  // TODO: Handle the case where there are no unassigned silos
  if (!unassignedSiloResult.data) {
    return null
  }

  // 2. Assign silo to a team
  const siloId = unassignedSiloResult.data.id

  await supabase
    .from("silos_teams")
    .insert([{ team_id: teamId, silo_id: siloId }])

  // 3. Update the silo based on the onboarding form details
  const siloUpdateData: Partial<Silo> = {}

  if (chainName) {
    siloUpdateData.name = chainName
  }

  const baseTokenConfig = BASE_TOKENS[baseToken]

  if (baseTokenConfig) {
    siloUpdateData.base_token_name = baseTokenConfig.name
    siloUpdateData.base_token_symbol = baseToken
  }

  const updateSiloResult = await supabase
    .from("silos")
    .update(siloUpdateData)
    .eq("id", siloId)
    .select()
    .single()

  assertNonNullSupabaseResult(updateSiloResult)
  assertValidSupabaseResult(updateSiloResult)

  return updateSiloResult.data
}

const updateSiloContract = async (silo: Silo) => {
  const siloDeployerApiClient = createSiloDeployerApiClient()

  // TODO: Place the transaction hash returned from this API call in a database,
  // so that we can track success or failure.
  if (silo.base_token_symbol) {
    await siloDeployerApiClient.setBaseToken({
      params: {
        contract_id: silo.engine_account,
      },
      data: {
        base_token_addr: silo.base_token_symbol,
      },
    })
  }
}

export const deploySilo = async (teamId: number) => {
  const onboardingForm = await getTeamOnboardingForm(teamId)

  if (!onboardingForm) {
    throw new Error(`Onboarding form not found for team ${teamId}`)
  }

  // Assign a silo to the team. At this point the silo contract will be
  // unconfigured (i.e. there is no base token). The row in the ACC database
  // should be seen as holding the desired, not necessarily the actual state.
  const silo = await assignSiloToTeam(teamId, onboardingForm)

  // TODO: Handle the case where no silo could not be assigned
  // (i.e. they have run out).
  if (!silo) {
    return
  }

  // Update the team to indicate that the deployment is in progress.
  await updateTeam(teamId, { onboarding_status: "DEPLOYMENT_IN_PROGRESS" })

  // Trigger a process to update the silo on the contract level.
  await updateSiloContract(silo)

  // For now, if the above call succeeds we will consider the deployment done.
  await updateTeam(teamId, { onboarding_status: "DEPLOYMENT_DONE" })
}
