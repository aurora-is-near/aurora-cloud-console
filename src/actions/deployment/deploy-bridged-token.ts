"use server"

import { JsonRpcProvider } from "ethers"
import { performSiloConfigTransaction } from "@/actions/deployment/perform-silo-config-transaction"
import { BridgedToken, Silo, SiloConfigTransactionStatus } from "@/types/types"
import { contractChangerApiClient } from "@/utils/contract-changer-api/contract-changer-api-client"
import { checkTokenByContractAddress } from "@/utils/check-token-contract"
import { getStorageBalanceByAddress } from "@/utils/near-storage"
import { BASE_TOKEN_PLACEHOLDER_ADDRESS } from "@/constants/base-token"
import { STORAGE_DEPOSIT_AMOUNT } from "@/constants/storage-deposits"

const checkContract = async ({
  provider,
  silo,
  bridgedToken,
  skipIfFailed,
}: {
  provider: JsonRpcProvider
  silo: Silo
  bridgedToken: BridgedToken
  skipIfFailed?: boolean
}): Promise<SiloConfigTransactionStatus> => {
  const nearAccountId = bridgedToken.near_address
  const siloAddress = bridgedToken.silo_address

  // Base tokens do not need an ERC20 contract deployed.
  if (
    silo.base_token_symbol.toUpperCase() === bridgedToken.symbol.toUpperCase()
  ) {
    return "SUCCESSFUL"
  }

  // If there is no silo address we can't mirror the token, or check if the
  // token contract was deployed. We return pending as the result for the case
  // where we fill this account ID in later.
  if (!siloAddress) {
    return "PENDING"
  }

  const isContractDeployed = await checkTokenByContractAddress(
    provider,
    siloAddress,
  )

  if (isContractDeployed) {
    return "SUCCESSFUL"
  }

  // If there is no near account ID is set we can't mirror the token. We return
  // pending as the result for the case where we fill this account ID in later.
  if (!nearAccountId) {
    return "PENDING"
  }

  return performSiloConfigTransaction(
    silo,
    "DEPLOY_TOKEN",
    async () =>
      contractChangerApiClient.mirrorErc20Token({
        siloEngineAccountId: silo.engine_account,
        token: {
          source_contract_id: "aurora",
          nep141: nearAccountId,
        },
      }),
    { skipIfFailed, nearAccountId },
  )
}

const checkStorageBalance = async ({
  silo,
  bridgedToken,
  skipIfFailed,
}: {
  silo: Silo
  bridgedToken: BridgedToken
  skipIfFailed?: boolean
}): Promise<SiloConfigTransactionStatus> => {
  const nearAccountId = bridgedToken.near_address

  // Covers the case where we are checking the base token (i.e. NEAR).
  if (nearAccountId === BASE_TOKEN_PLACEHOLDER_ADDRESS) {
    return "SUCCESSFUL"
  }

  // If no near account ID is set we can't check the storage balance. We return
  // pending as the result for the case where we fill this account ID in later.
  if (!nearAccountId) {
    return "PENDING"
  }

  const storageBalance = await getStorageBalanceByAddress(
    silo.engine_account,
    nearAccountId,
  )

  if (storageBalance?.total) {
    return "SUCCESSFUL"
  }

  return performSiloConfigTransaction(
    silo,
    "STORAGE_DEPOSIT",
    async () =>
      contractChangerApiClient.makeStorageDeposit({
        siloEngineAccountId: silo.engine_account,
        amount: STORAGE_DEPOSIT_AMOUNT,
        token: nearAccountId,
      }),
    { skipIfFailed, nearAccountId },
  )
}

export const deployBridgedToken = async ({
  silo,
  bridgedToken,
  skipIfFailed,
}: {
  silo: Silo
  bridgedToken: BridgedToken
  skipIfFailed?: boolean
}): Promise<SiloConfigTransactionStatus> => {
  const provider = new JsonRpcProvider(silo.rpc_url)

  const statuses = await Promise.all([
    checkContract({
      provider,
      silo,
      bridgedToken,
      skipIfFailed,
    }),
    checkStorageBalance({ silo, bridgedToken, skipIfFailed }),
  ])

  if (statuses.includes("PENDING")) {
    return "PENDING"
  }

  if (statuses.includes("FAILED")) {
    return "FAILED"
  }

  return "SUCCESSFUL"
}
