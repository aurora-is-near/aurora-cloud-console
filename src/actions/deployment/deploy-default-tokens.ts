"use server"

import { JsonRpcProvider } from "ethers"
import { performSiloConfigTransaction } from "@/actions/deployment/perform-silo-config-transaction"
import {
  Silo,
  SiloConfigTransactionOperation,
  SiloConfigTransactionStatus,
} from "@/types/types"
import { contractChangerApiClient } from "@/utils/contract-changer-api/contract-changer-api-client"
import { checkTokenBySymbol } from "@/utils/check-token-contract"
import { DefaultToken } from "@/types/default-tokens"
import { DEFAULT_TOKENS } from "@/constants/default-tokens"
import { getStorageBalanceBySymbol } from "@/utils/near-storage"
import { NEAR_TOKEN_ADDRESSES } from "@/constants/near-token"
import { STORAGE_DEPOSIT_AMOUNT } from "@/constants/storage-deposits"

const CONTRACT_CHANGER_SYMBOLS: Record<
  DefaultToken,
  Parameters<typeof contractChangerApiClient.mirrorErc20Token>[0]["token"]
> = {
  NEAR: "Near",
  AURORA: "Aurora",
  USDt: "Usdt",
  USDC: "Usdc",
  ETH: {
    source_contract_id: "0x4e454160.c.aurora",
    nep141: "eth.bridge.near",
  },
}

const SILO_CONFIG_TRANSACTION_OPERATIONS: Record<
  DefaultToken,
  SiloConfigTransactionOperation
> = {
  NEAR: "DEPLOY_NEAR",
  AURORA: "DEPLOY_AURORA",
  USDt: "DEPLOY_USDT",
  USDC: "DEPLOY_USDC",
  ETH: "DEPLOY_ETH",
}

const checkContract = async ({
  provider,
  silo,
  symbol,
  skipIfFailed,
}: {
  provider: JsonRpcProvider
  silo: Silo
  symbol: DefaultToken
  skipIfFailed?: boolean
}) => {
  const isContractDeployed = await checkTokenBySymbol(provider, symbol)
  const nearAccountId = symbol === "NEAR" ? null : NEAR_TOKEN_ADDRESSES[symbol]

  if (isContractDeployed) {
    return "SUCCESSFUL"
  }

  return performSiloConfigTransaction(
    silo,
    SILO_CONFIG_TRANSACTION_OPERATIONS[symbol],
    async () =>
      contractChangerApiClient.mirrorErc20Token({
        siloEngineAccountId: silo.engine_account,
        token: CONTRACT_CHANGER_SYMBOLS[symbol],
      }),
    { skipIfFailed, nearAccountId },
  )
}

const checkStorageBalance = async ({
  silo,
  symbol,
  skipIfFailed,
}: {
  silo: Silo
  symbol: DefaultToken
  skipIfFailed?: boolean
}) => {
  if (symbol === "NEAR") {
    return "SUCCESSFUL"
  }

  const storageBalance = await getStorageBalanceBySymbol(
    silo.engine_account,
    symbol,
  )

  if (storageBalance?.total) {
    return "SUCCESSFUL"
  }

  const nearAccountId = NEAR_TOKEN_ADDRESSES[symbol]

  return performSiloConfigTransaction(
    silo,
    "STORAGE_DEPOSIT",
    async () =>
      contractChangerApiClient.makeStorageDeposit({
        siloEngineAccountId: silo.engine_account,
        amount: STORAGE_DEPOSIT_AMOUNT,
        token: nearAccountId,
      }),
    { skipIfFailed, nearAccountId },
  )
}

const deployDefaultToken = async ({
  provider,
  silo,
  symbol,
  skipIfFailed,
}: {
  provider: JsonRpcProvider
  silo: Silo
  symbol: DefaultToken
  skipIfFailed?: boolean
}): Promise<SiloConfigTransactionStatus[]> => {
  return Promise.all([
    checkContract({ provider, silo, symbol, skipIfFailed }),
    checkStorageBalance({ silo, symbol, skipIfFailed }),
  ])
}

export const deployDefaultTokens = async (
  silo: Silo,
  { skipIfFailed }: { skipIfFailed?: boolean } = {},
): Promise<SiloConfigTransactionStatus> => {
  const provider = new JsonRpcProvider(silo.rpc_url)

  const statuses = await DEFAULT_TOKENS.filter(
    (token) => token !== silo.base_token_symbol,
  ).reduce(
    async (acc, symbol) => {
      const previousStatuses = await acc
      const symbolStatuses = await deployDefaultToken({
        provider,
        silo,
        symbol,
        skipIfFailed,
      })

      return [...previousStatuses, ...symbolStatuses]
    },
    Promise.resolve([] as SiloConfigTransactionStatus[]),
  )

  if (statuses.includes("PENDING")) {
    return "PENDING"
  }

  if (statuses.includes("FAILED")) {
    return "FAILED"
  }

  return "SUCCESSFUL"
}
