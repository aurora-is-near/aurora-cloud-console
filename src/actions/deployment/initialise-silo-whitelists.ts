"use server"

import { performSiloConfigTransaction } from "@/actions/deployment/perform-silo-config-transaction"
import { Silo, SiloConfigTransactionStatus } from "@/types/types"
import { contractChangerApiClient } from "@/utils/contract-changer-api/contract-changer-api-client"

export const initialiseSiloWhitelists = async (
  silo: Silo,
  {
    skipIfFailed,
  }: {
    skipIfFailed?: boolean
  } = {},
): Promise<SiloConfigTransactionStatus> => {
  const promises = []

  if (!silo.is_make_txs_public) {
    promises.push(
      performSiloConfigTransaction(
        silo,
        "ENABLE_MAKE_TXS_WHITELIST",
        async () => {
          return contractChangerApiClient.toggleWhitelist({
            siloEngineAccountId: silo.engine_account,
            action: "disable",
            whitelistKind: "address",
          })
        },
        { skipIfFailed, nearAccountId: silo.engine_account },
      ),
    )
  }

  if (!silo.is_deploy_contracts_public) {
    promises.push(
      performSiloConfigTransaction(
        silo,
        "ENABLE_DEPLOY_CONTRACT_WHITELIST",
        async () => {
          return contractChangerApiClient.toggleWhitelist({
            siloEngineAccountId: silo.engine_account,
            action: "disable",
            whitelistKind: "evm-admin",
          })
        },
        { skipIfFailed, nearAccountId: silo.engine_account },
      ),
    )
  }

  const [statuses] = await Promise.all(promises)

  if (statuses.includes("PENDING")) {
    return "PENDING"
  }

  if (statuses.includes("FAILED")) {
    return "FAILED"
  }

  return "SUCCESSFUL"
}
