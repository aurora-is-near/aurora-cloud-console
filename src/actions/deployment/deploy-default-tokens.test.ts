import { ethers } from "ethers"
import { symbol } from "zod"
import { contractChangerApiClient } from "@/utils/contract-changer-api/contract-changer-api-client"
import { deployDefaultTokens } from "@/actions/deployment/deploy-default-tokens"
import { createMockSilo } from "../../../test-utils/factories/silo-factory"
import {
  createSelect,
  mockSupabaseClient,
} from "../../../test-utils/mock-supabase-client"

jest.mock("ethers")
jest.mock("@/utils/contract-changer-api/contract-changer-api-client", () => ({
  contractChangerApiClient: {
    mirrorErc20Token: jest.fn(({ token }) => ({
      tx_hash: `mock_tx_hash_${token}`,
    })),
  },
}))

const mockTxStatus = jest.fn()

jest.mock("near-api-js", () => ({
  ...jest.requireActual("near-api-js"),
  connect: jest.fn(() => ({
    connection: {
      provider: {
        txStatus: mockTxStatus,
      },
    },
  })),
}))

const mockSilo = createMockSilo({ base_token_symbol: "AURORA" })
const contractChangerTokens = ["Near", "Aurora", "Usdt", "Usdc"]
const nonBaseTokens = contractChangerTokens.filter(
  (token) => token !== "Aurora",
)

describe("deployDefaultTokens", () => {
  beforeEach(() => {
    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockReturnValue(createSelect([]))
    ;(ethers.Contract as jest.Mock).mockImplementation(() => ({
      symbol: () => {
        throw new Error("Not implemented")
      },
    }))
  })

  it("performs the transactions to deploy the tokens", async () => {
    const result = await deployDefaultTokens(mockSilo)

    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenCalledTimes(
      nonBaseTokens.length,
    )

    nonBaseTokens.forEach((token) => {
      expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenCalledWith({
        siloEngineAccountId: mockSilo.engine_account,
        token,
      })
    })

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledTimes(3)

    nonBaseTokens.forEach((token) => {
      expect(
        mockSupabaseClient.from("silo_config_transactions").insert,
      ).toHaveBeenCalledWith({
        silo_id: mockSilo.id,
        operation: `DEPLOY_${token.toUpperCase()}`,
        status: "PENDING",
        transaction_hash: `mock_tx_hash_${token}`,
      })
    })

    expect(result).toBe("PENDING")
  })

  it("skips if all transactions were already SUCCESSFUL", async () => {
    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockReturnValue(createSelect([{ status: "SUCCESSFUL" }]))

    const result = await deployDefaultTokens(mockSilo)

    expect(contractChangerApiClient.mirrorErc20Token).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).not.toHaveBeenCalled()
    expect(mockTxStatus).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").update,
    ).not.toHaveBeenCalled()

    expect(result).toBe("SUCCESSFUL")
  })

  it("skips a single token with an already SUCCESSFUL transaction", async () => {
    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockReturnValueOnce(createSelect([{ status: "SUCCESSFUL" }]))

    const result = await deployDefaultTokens(mockSilo)

    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenCalledTimes(
      nonBaseTokens.length - 1,
    )

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledTimes(nonBaseTokens.length - 1)

    nonBaseTokens.slice(1).forEach((token, i) => {
      expect(
        mockSupabaseClient.from("silo_config_transactions").insert,
      ).toHaveBeenNthCalledWith(i + 1, {
        silo_id: mockSilo.id,
        operation: `DEPLOY_${token.toUpperCase()}`,
        status: "PENDING",
        transaction_hash: `mock_tx_hash_${token}`,
      })
    })

    expect(mockTxStatus).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").update,
    ).not.toHaveBeenCalled()

    expect(result).toBe("PENDING")
  })

  it("skips if a token contract was already deployed", async () => {
    ;(ethers.Contract as jest.Mock).mockImplementation(
      (tokenContractAddress) => ({
        symbol: () => {
          if (
            tokenContractAddress ===
            "0x80Da25Da4D783E57d2FCdA0436873A193a4BEccF"
          ) {
            return "USDt"
          }

          throw new Error("Not implemented")
        },
      }),
    )

    const result = await deployDefaultTokens(mockSilo)

    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenCalledTimes(2)
    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenCalledWith({
      siloEngineAccountId: mockSilo.engine_account,
      token: "Near",
    })

    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenCalledWith({
      siloEngineAccountId: mockSilo.engine_account,
      token: "Usdc",
    })

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledTimes(2)

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledWith({
      operation: "DEPLOY_NEAR",
      silo_id: 1,
      status: "PENDING",
      transaction_hash: "mock_tx_hash_Near",
    })

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledWith({
      operation: "DEPLOY_USDC",
      silo_id: 1,
      status: "PENDING",
      transaction_hash: "mock_tx_hash_Usdc",
    })

    expect(mockTxStatus).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").update,
    ).not.toHaveBeenCalled()

    expect(result).toBe("PENDING")
  })

  it("handles FAILED transactions", async () => {
    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockReturnValue(createSelect([{ status: "FAILED" }]))

    const result = await deployDefaultTokens(mockSilo)

    expect(contractChangerApiClient.mirrorErc20Token).toHaveBeenCalledTimes(
      nonBaseTokens.length,
    )

    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).toHaveBeenCalledTimes(nonBaseTokens.length)

    nonBaseTokens.forEach((token) => {
      expect(
        mockSupabaseClient.from("silo_config_transactions").insert,
      ).toHaveBeenCalledWith({
        silo_id: mockSilo.id,
        operation: `DEPLOY_${token.toUpperCase()}`,
        status: "PENDING",
        transaction_hash: `mock_tx_hash_${token}`,
      })
    })

    expect(mockTxStatus).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").update,
    ).not.toHaveBeenCalled()

    expect(result).toBe("PENDING")
  })

  it("handles transactions that are still PENDING", async () => {
    let calls = 0

    mockTxStatus.mockResolvedValue({ status: "NotStarted" })

    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockImplementation(() => {
        calls += 1

        return createSelect([
          {
            status: "PENDING",
            transaction_hash: `mock_tx_hash_${calls}`,
          },
        ])
      })

    const result = await deployDefaultTokens(mockSilo)

    expect(contractChangerApiClient.mirrorErc20Token).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).not.toHaveBeenCalled()

    expect(mockTxStatus).toHaveBeenCalledTimes(nonBaseTokens.length)

    Array.from({ length: nonBaseTokens.length }).forEach((_, i) => {
      expect(mockTxStatus).toHaveBeenNthCalledWith(
        i + 1,
        `mock_tx_hash_${i + 1}`,
        mockSilo.engine_account,
        "FINAL",
      )
    })

    expect(
      mockSupabaseClient.from("silo_config_transactions").update,
    ).not.toHaveBeenCalled()

    expect(result).toBe("PENDING")
  })

  it("handles PENDING transactions that has since been successful", async () => {
    let calls = 0

    mockTxStatus.mockResolvedValue({ status: { SuccessValue: "" } })

    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockImplementation(() => {
        calls += 1

        return createSelect([
          {
            status: "PENDING",
            transaction_hash: `mock_tx_hash_${calls}`,
          },
        ])
      })

    const result = await deployDefaultTokens(mockSilo)

    expect(contractChangerApiClient.mirrorErc20Token).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).not.toHaveBeenCalled()

    expect(mockTxStatus).toHaveBeenCalledTimes(nonBaseTokens.length)

    Array.from({ length: nonBaseTokens.length }).forEach((_, i) => {
      expect(mockTxStatus).toHaveBeenNthCalledWith(
        i + 1,
        `mock_tx_hash_${i + 1}`,
        mockSilo.engine_account,
        "FINAL",
      )
    })

    expect(
      mockSupabaseClient.from("silo_config_transactions").update,
    ).toHaveBeenCalledTimes(nonBaseTokens.length)

    Array.from({ length: nonBaseTokens.length }).forEach((_, i) => {
      expect(
        mockSupabaseClient.from("silo_config_transactions").update,
      ).toHaveBeenNthCalledWith(i + 1, { status: "SUCCESSFUL" })
    })

    expect(result).toBe("SUCCESSFUL")
  })

  it("handles PENDING transactions that have since failed", async () => {
    let calls = 0

    mockTxStatus.mockResolvedValue({ status: { Failure: {} } })

    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockImplementation(() => {
        calls += 1

        return createSelect([
          {
            status: "PENDING",
            transaction_hash: `mock_tx_hash_${calls}`,
          },
        ])
      })

    const result = await deployDefaultTokens(mockSilo)

    expect(contractChangerApiClient.mirrorErc20Token).not.toHaveBeenCalled()
    expect(
      mockSupabaseClient.from("silo_config_transactions").insert,
    ).not.toHaveBeenCalled()

    expect(mockTxStatus).toHaveBeenCalledTimes(nonBaseTokens.length)

    Array.from({ length: nonBaseTokens.length }).forEach((_, i) => {
      expect(mockTxStatus).toHaveBeenNthCalledWith(
        i + 1,
        `mock_tx_hash_${i + 1}`,
        mockSilo.engine_account,
        "FINAL",
      )
    })

    expect(
      mockSupabaseClient.from("silo_config_transactions").update,
    ).toHaveBeenCalledTimes(nonBaseTokens.length)

    Array.from({ length: nonBaseTokens.length }).forEach((_, i) => {
      expect(
        mockSupabaseClient.from("silo_config_transactions").update,
      ).toHaveBeenNthCalledWith(i + 1, { status: "FAILED" })
    })

    expect(result).toBe("FAILED")
  })
})
