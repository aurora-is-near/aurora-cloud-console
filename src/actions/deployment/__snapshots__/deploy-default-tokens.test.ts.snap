// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deployDefaultTokens handles FAILED transactions 1`] = `
[
  [
    {
      "operation": "DEPLOY_NEAR",
      "silo_id": 1,
      "status": "PENDING",
      "target": null,
      "transaction_hash": "mock_tx_hash_Near",
    },
  ],
  [
    {
      "operation": "DEPLOY_USDT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "usdt.tether-token.near",
      "transaction_hash": "mock_tx_hash_Usdt",
    },
  ],
  [
    {
      "operation": "DEPLOY_USDC",
      "silo_id": 1,
      "status": "PENDING",
      "target": "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
      "transaction_hash": "mock_tx_hash_Usdc",
    },
  ],
  [
    {
      "operation": "DEPLOY_ETH",
      "silo_id": 1,
      "status": "PENDING",
      "target": "eth.bridge.near",
      "transaction_hash": "mock_tx_hash_[object Object]",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "usdt.tether-token.near",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "eth.bridge.near",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
]
`;

exports[`deployDefaultTokens performs the transactions to deploy the tokens 1`] = `
[
  [
    {
      "operation": "DEPLOY_NEAR",
      "silo_id": 1,
      "status": "PENDING",
      "target": null,
      "transaction_hash": "mock_tx_hash_Near",
    },
  ],
  [
    {
      "operation": "DEPLOY_USDT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "usdt.tether-token.near",
      "transaction_hash": "mock_tx_hash_Usdt",
    },
  ],
  [
    {
      "operation": "DEPLOY_USDC",
      "silo_id": 1,
      "status": "PENDING",
      "target": "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
      "transaction_hash": "mock_tx_hash_Usdc",
    },
  ],
  [
    {
      "operation": "DEPLOY_ETH",
      "silo_id": 1,
      "status": "PENDING",
      "target": "eth.bridge.near",
      "transaction_hash": "mock_tx_hash_[object Object]",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "usdt.tether-token.near",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "eth.bridge.near",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
]
`;

exports[`deployDefaultTokens skips a single token with an already SUCCESSFUL transaction 1`] = `
[
  [
    {
      "operation": "DEPLOY_USDT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "usdt.tether-token.near",
      "transaction_hash": "mock_tx_hash_Usdt",
    },
  ],
  [
    {
      "operation": "DEPLOY_USDC",
      "silo_id": 1,
      "status": "PENDING",
      "target": "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
      "transaction_hash": "mock_tx_hash_Usdc",
    },
  ],
  [
    {
      "operation": "DEPLOY_ETH",
      "silo_id": 1,
      "status": "PENDING",
      "target": "eth.bridge.near",
      "transaction_hash": "mock_tx_hash_[object Object]",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "usdt.tether-token.near",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "eth.bridge.near",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
]
`;

exports[`deployDefaultTokens skips if a token contract was already deployed 1`] = `
[
  [
    {
      "operation": "DEPLOY_NEAR",
      "silo_id": 1,
      "status": "PENDING",
      "target": null,
      "transaction_hash": "mock_tx_hash_Near",
    },
  ],
  [
    {
      "operation": "DEPLOY_USDC",
      "silo_id": 1,
      "status": "PENDING",
      "target": "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
      "transaction_hash": "mock_tx_hash_Usdc",
    },
  ],
  [
    {
      "operation": "DEPLOY_ETH",
      "silo_id": 1,
      "status": "PENDING",
      "target": "eth.bridge.near",
      "transaction_hash": "mock_tx_hash_[object Object]",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "usdt.tether-token.near",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
  [
    {
      "operation": "STORAGE_DEPOSIT",
      "silo_id": 1,
      "status": "PENDING",
      "target": "eth.bridge.near",
      "transaction_hash": "mock_storage_deposit_tx_hash",
    },
  ],
]
`;
